### PR TITLE
Admin API DNS Rebinding mitigation

### PR DESCRIPTION
**Pull request by DerTiedemann**
New PR as #231 was merged to the wrong branch

## Why this should be merged
To fix a DNS rebinding attack against locally running node.
Camino node exposes the administrator functionality via /ext/admin RPC API endpoint. For example, it is possible to retrieve a private key of the node by sending the following payload:
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method" : "admin.getNodeSigner"
}
```

While /ext/admin is disabled on public nodes, it is still possible to access
administrator endpoints from the internet via DNS rebinding when a node is
run locally. To do so, an attacker lures a victim with a running local Camino
node on an HTML page with a specially crafted JavaScript scenario. After
the victim visits the malicious page, A record in DNS settings is
automatically changed to resolve to the locally running node (127.0.0.1),
this page periodically sends admin.getNodeSigner HTTP requests to the
local node of the victim, and after some time the browser receives an
updated DNS information and starts to resolve a malicious domain to the
local node service thus bypassing Same Origin Policy. For example, in Safari
it takes roughly 10 minutes for the browser to fetch new DNS settings. In
Chrome there is a protection from this attack that explicitly forbids 127.0.0.1,
however it can be bypassed by setting 0.0.0.0 in the A record.
## How this works
This assumes the attacker has no control over the Host header of a sent address. (This could not be the case, please verify)
Now every request to the admin API needs to have the hostname that ist set in the `--http-host` flag. Otherwise requests will fail.
This is not optional.
## How this was tested
Added two new tests.

### CODE REVIEW

File: api/admin/service.go [hunk 1]
```diff
@@ -83,9 +89,48 @@ func NewService(config Config) (*common.HTTPHandler, error) {
...
+		return nil, err
+	}
+
+	newServer.RegisterValidateRequestFunc(mw)
+
 	return &common.HTTPHandler{Handler: newServer}, nil
 }
 
+func BlockRequestsWithNonMatchingHostnames(hostName string) (func(r *rpc.RequestInfo, _i interface{}) error, error) {
+	host := strings.Split(hostName, ":")
```
Comment (knikos) [inline api/admin/service.go:107]: the configured Host `config.HTTPHost` cannot include a port afaik. The request host `r.Request.Host` is the one that can contain a port.

Comment (DerTiedemann) [reply api/admin/service.go:107]: thats fine - the we will always hit the second case
this is here for completeness sake and if we decided to pass the port

File: node/config.go [hunk 2]
```diff
@@ -65,11 +65,12 @@ type APIConfig struct {
 	IPCConfig        `json:"ipcConfig"`
 
 	// Enable/Disable APIs
-	AdminAPIEnabled    bool `json:"adminAPIEnabled"`
-	InfoAPIEnabled     bool `json:"infoAPIEnabled"`
-	KeystoreAPIEnabled bool `json:"keystoreAPIEnabled"`
-	MetricsAPIEnabled  bool `json:"metricsAPIEnabled"`
-	HealthAPIEnabled   bool `json:"healthAPIEnabled"`
+	AdminAPIEnabled       bool `json:"adminAPIEnabled"`
+	UnsafeAdminAPIEnabled bool `json:"unsageAdminAPIEnabled"`
```
Comment (knikos) [inline node/config.go]: minor typo here: unsa**ge**AdminAPIEnabled

**Comment by knikos** [comment]
the flagset is missing the new Flag key --unsafe-api-admin-enabled

Something like the following should be added in `config/flags.go#addNodeFlags`

`fs.Bool(AdminAPIDisableSafeModeKey, false, "If true, this node exposes the Admin API without safe mode")`
Otherwise it prints the error
`couldn't configure flags: unknown flag: --unsafe-api-admin-enabled`

**Review by knikos — CHANGES_REQUESTED** [review]

**Comment by DerTiedemann** [comment]
> the flagset is missing the new Flag key --unsafe-api-admin-enabled
> 
> Something like the following should be added in `config/flags.go#addNodeFlags`
> 
> `fs.Bool(AdminAPIDisableSafeModeKey, false, "If true, this node exposes the Admin API without safe mode")` Otherwise it prints the error `couldn't configure flags: unknown flag: --unsafe-api-admin-enabled`

Resolved

**Review by knikos — APPROVED** [review]

**Pull request merged**